tiledb/common/lru.h

```diff
@@ -0,0 +1,76 @@
+template <class Key, class Value>
+class QueryCache {
+ public:
+  using List = std::list<Key>;
+  using OndemandValue = std::promise<Value>;
+  using Cached_Entry = std::tuple<List::iterator, OndemandValue>;
+
+  /** Constructor */
+  QueryCache(size_t memory_budget);
+
+  /**
+   * Try to get an entry from the cache given a key.
+   * The function returns std::nullopt the the key is not present
+   * in the cache or a promised value which the caller can
+   * sync() on to fetch it or return from memory.
+   * If the entry exists, the get() operation moves this entry
+   * at the end of the internal LRU list.
+   */
+  std::optional<OndemandValue> get(const Key& key);
+
+  /**
+   * Returns the value from a cached entry.
+   * This gets a std::shared_future from the promised value and
+   * wait()s on it for Value to become available. The wait() call
+   * might block until the other reader fetches the value or will
+   * return immediately if the value is ready in the cache.
+   */
+  Value sync(OndemandValue promised_val);
+
+  /**
+   * Inserts an entry into the cache, executes the callback to
+   * fetch the value and sets it on the promise so that all the
+   * other concurrent readers waiting on the future to get unblocked.
+   * If the key already exists in the cache, exit early. Even if the
+   * existing entry is inprogress, it'd be pointless to start fetching
+   * the value here again.
+   * TODO: A put() operation will add to memory_consumed_, signatures
+   * might change here to allow access to how much memory the Value consumes.
+   */
+  void put(const Key& key, std::function<Value()> cb);
+
+ private:
+  /**
+   * Pops the head of the LRU list.
+   *
+   * TODO: If the evicted entry has an inprogress promise, it's tricky.
+   * V1: we wait for the promise to be ready, thus we make sure the total
+   * amount of RAM stays within memory_budget_.
+   * V2: We just evict, let the memory niddle slide a bit
+   * past the red line and readers will sync with each other peer-to-peer.
+   */
+  void evict();
+
+  /**
+   * Double linked list where the head is the least
+   * recently used element from the cache.
+   */
+  List lru_;
+
+  /**
+   * Map of cached entries by key. Each entry stores and iterator
+   * within the LRU list and a promised value which is either already
+   * available to serve readers or it is being fetched from storage and
+   * readers can wait for it to become available
+   */
+  std::unordered_map<Key, Cached_Entry> entries_;
```

Comment:
One of these containers needs to designated as the owner of cache entries and the other as subordinate to it. Otherwise there are going to be lifespan issues.